deployments/extensions.go

```diff
@@ -0,0 +1,405 @@
+package deployments
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	"github.com/fuseml/fuseml/cli/helpers"
+	"github.com/fuseml/fuseml/cli/kubernetes"
+	"github.com/fuseml/fuseml/cli/paas/ui"
+)
+
+const (
+	defaultDescriptionFileName = "description.yaml"
+	defaultNamespace           = "fuseml-workloads"
+)
+
+type installStep struct {
+	Type      string
+	Location  string
+	Values    string
+	Namespace string
+}
+
+type istioGateway struct {
+	Name        string
+	Port        int
+	ServiceHost string
+}
+
+type extensionDesc struct {
+	Name        string
+	Description string
+	Namespace   string
+	Install     []installStep
+	Uninstall   []installStep
+	Gateways    []istioGateway
+}
+
+type Extension struct {
+	Name       string
+	Repository string
+	Debug      bool
+	desc       *extensionDesc
+}
+
+func NewExtension(name, repository string) *Extension {
+	return &Extension{
+		Name:       name,
+		Repository: repository,
+		desc:       &extensionDesc{},
+		Debug:      false,
+	}
+}
+
+// LoadDescription finds the description file of the extension and loads it into the struct
+func (e *Extension) LoadDescription() error {
+
+	u, err := url.Parse(e.Repository)
+	if err != nil {
+		return err
+	}
+
+	tmpDir, err := ioutil.TempDir("", "fuseml-extension")
+	if err != nil {
+		return errors.Wrap(err, "can't create temp directory "+tmpDir)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	descFilePath := ""
+
+	if u.IsAbs() && u.Scheme != "" && u.Host != "" {
+		// "/" at the end is necessary so that last part of the path is not replaced
+		u, _ = u.Parse(e.Name + "/")
+		u, _ = u.Parse(defaultDescriptionFileName)
+		if err := helpers.DownloadFile(u.String(), defaultDescriptionFileName, tmpDir); err != nil {
+			return err
+		}
+		descFilePath = filepath.Join(tmpDir, defaultDescriptionFileName)
+	} else {
+		info, err := os.Stat(e.Repository)
+		if os.IsNotExist(err) {
+			return err
+		}
+		if !info.IsDir() {
+			return errors.New("Provided path to extension repository is neither URL nor a directory")
+
+		}
+		descFilePath = filepath.Join(e.Repository, e.Name, defaultDescriptionFileName)
+	}
+
+	// parse and load descriptin file into Extension struct
+	data, err := os.ReadFile(descFilePath)
+	if err != nil {
+		return errors.Wrap(err, "failed to read description file")
+	}
+
+	err = yaml.Unmarshal(data, &e.desc)
```

Comment:
could you give a try to this yaml package: `github.com/ghodss/yaml` ?


This is what we are using in fuseml/fuseml-core, after we switched from gopkg.in/yaml.v2 due to its issues with capital characters. I am pretty sure you will not have the same problem here, however if one day we move fuseml-core to fuseml at least we would have been using the same yaml library.

Reply:
I'll try